src/db_adapters/assets/non_fungible_token_events.rs

```diff
@@ -0,0 +1,202 @@
+            outcome,
+            block_timestamp,
+            &shard.shard_id,
+            &mut index_in_shard,
+        );
+
+        crate::await_retry_or_panic!(
+            diesel::insert_into(schema::assets__non_fungible_token_events::table)
+                .values(nft_events.clone())
+                .execute_async(pool),
+            10,
+            "NonFungibleTokenEvent were adding to database".to_string(),
+            &nft_events,
+            &is_error_handled
+        );
+    }
+    Ok(())
+}
+
+async fn is_error_handled(async_error: &AsyncError<diesel::result::Error>) -> bool {
+    if let actix_diesel::AsyncError::Execute(diesel::result::Error::DatabaseError(
+        diesel::result::DatabaseErrorKind::UniqueViolation,
+        ref error_info,
+    )) = *async_error
+    {
+        let duplicate_constraint = "assets__non_fungible_token_events_pkey";
+        let broken_data_constraint = "assets__non_fungible_token_events_unique";
+        let constraint_name = error_info.constraint_name().unwrap_or("");
+        if constraint_name == duplicate_constraint {
+            // Everything is fine, we have already written this to the DB
+            return true;
+        } else if constraint_name == broken_data_constraint {
+            warn!(
+                target: crate::INDEXER_FOR_EXPLORER,
+                "NFT: data inconsistency is found"
+            );
+        }
+    }
+    false
+}
+
+fn collect_nft_events(
+    outcome: &near_indexer::IndexerExecutionOutcomeWithReceipt,
+    block_timestamp: &u64,
+    shard_id: &near_indexer::near_primitives::types::ShardId,
+    index_in_shard: &mut i32,
+) -> Vec<models::assets::non_fungible_token_events::NonFungibleTokenEvent> {
+    let prefix = "EVENT_JSON:";
+    let event_logs: Vec<nft_types::Nep171Event> = outcome.execution_outcome.outcome.logs.iter().filter_map(|untrimmed_log| {
+        // Now we have only nep171 events, we both parse the logs and handle nep171 here.
+        // When other event types will be added, we need to rewrite the logic
+        // so that we parse the logs only once for all,
+        // and then handle them for each event type separately.
+        let log = untrimmed_log.trim();
+        if !log.starts_with(prefix) {
+            return None;
+        }
+
+        let event: nft_types::NearEvent = match serde_json::from_str::<'_, nft_types::NearEvent>(
+            log[prefix.len()..].trim(),
+        ) {
+            Ok(result) => result,
+            Err(err) => {
+                warn!(
+                    target: crate::INDEXER_FOR_EXPLORER,
+                    "NFT: provided event log does not correspond to any of formats defined in NEP. Will ignore this event. \n {:#?} \n{:#?}",
+                    err,
+                    untrimmed_log,
+                );
+                return None;
+            }
+        };
+
+        let nft_types::NearEvent::Nep171(nep171_event) = event;
+        Some(nep171_event)
+    }).collect();
```

Comment:
Just FYI, my point was to avoid Vec. You should be able to just drop .collect() (and drop the explicit type) and that will give you an iterator, so you can directly use it in the loop

Reply:
Oh, now I see.
I was thinking about `map` function in the next piece of code, but the problem was that each event can produce multiple DB rows, and I don't know how to write it in a functional style in Rust.

I was happy to change at least the first cycle here, but I didn't finish it properly 😄 
I will fix it in another small PR!